### PR TITLE
.basis file support

### PR DESCRIPTION
**Pull request by TrevorDev**
https://github.com/BabylonJS/Babylon.js/issues/6382

TestPG: #4RN0VF

image

### CODE REVIEW

File: src/Materials/Textures/Loaders/basisTextureLoader.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Upload data to texture
+        callback(fileInfo.width, fileInfo.height, false, true, () => {
+            if (transcodeResult.fallbackToRgb565) {
+                // Load rgb565Data to texture if conversion is needed
+                texture.getEngine()._unpackFlipY(texture.invertY);
+                var gl = texture.getEngine()._gl;
+                gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGB, fileInfo.alignedWidth, fileInfo.alignedHeight, 0, gl.RGB, gl.UNSIGNED_SHORT_5_6_5, transcodeResult.pixels);
+
+                // TODO this is not working because computed power of 2 image size is larger than rgb565Data causing an error, working around this with below code
```
Comment (TrevorDev) [inline src/Materials/Textures/Loaders/basisTextureLoader.ts]: Any ideas on how to support this?

File: Playground/index-local.html [hunk 2]
```diff
@@ -22,6 +22,7 @@
     <script src="../dist/preview%20release/Oimo.js"></script>
     <script src="../dist/preview%20release/gltf_validator.js"></script>
     <script src="../dist/preview%20release/earcut.min.js"></script>
+    <script src="../dist/preview%20release/basisTranscoder/basis_transcoder.js"></script>
```
Comment (deltakosh) [inline Playground/index-local.html]: we need to do it for all html files (full, frame, index, etc)

File: src/Materials/Textures/Loaders/basisTextureLoader.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import { Engine } from "../../../Engines/engine";
+import { InternalTexture } from "../../../Materials/Textures/internalTexture";
+import { IInternalTextureLoader } from "../../../Materials/Textures/internalTextureLoader";
+import { _TimeToken } from "../../../Instrumentation/timeToken";
+import { _DepthCullingState, _StencilState, _AlphaState } from "../../../States/index";
+import { BasisTools } from "../../../Misc/basis";
+
+export class _BasisTextureLoader implements IInternalTextureLoader {
+    /**
+     * Defines wether the loader supports cascade loading the different faces.
```
Comment (deltakosh) [inline src/Materials/Textures/Loaders/basisTextureLoader.ts]: typo

File: src/Materials/Textures/Loaders/basisTextureLoader.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+     * This returns if the loader support the current file information.
+     * @param extension defines the file extension of the file being loaded
+     * @param textureFormatInUse defines the current compressed format in use iun the engine
+     * @param fallback defines the fallback internal texture if any
+     * @param isBase64 defines whether the texture is encoded as a base64
+     * @param isBuffer defines whether the texture data are stored as a buffer
+     * @returns true if the loader can load the specified file
+     */
+    public canLoad(extension: string, textureFormatInUse: Nullable<string>, fallback: Nullable<InternalTexture>, isBase64: boolean, isBuffer: boolean): boolean {
+        console.log("basis try");
```
Comment (deltakosh) [inline src/Materials/Textures/Loaders/basisTextureLoader.ts]: console.log

File: src/Materials/Textures/Loaders/basisTextureLoader.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+     * @param rootUrl the url of the texture
+     * @param textureFormatInUse defines the current compressed format in use iun the engine
+     * @returns the fallback texture
+     */
+    public getFallbackTextureUrl(rootUrl: string, textureFormatInUse: Nullable<string>): Nullable<string> {
+        return null;
+    }
+
+    /**
+     * Uploads the cube texture data to the WebGl Texture. It has alreday been bound.
```
Comment (deltakosh) [inline src/Materials/Textures/Loaders/basisTextureLoader.ts]: typo

File: src/Materials/Textures/Loaders/basisTextureLoader.ts (hunk 1 above)
Comment (deltakosh) [reply src/Materials/Textures/Loaders/basisTextureLoader.ts]: Not sure to understand why it is different from a regular texture at that point? We do the same rescale for regular jpg texture for instance when we need pow 2

Comment (deltakosh) [reply src/Materials/Textures/Loaders/basisTextureLoader.ts]: like here: https://github.com/BabylonJS/Babylon.js/blob/master/src/Engines/engine.ts#L4312

File: src/Misc/basis.ts [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+        }else if (caps.pvrtc) {
+            format = BasisTools._BASIS_FORMAT.cTFPVRTC1_4_OPAQUE_ONLY;
+        }else if (caps.etc2) {
+            format = BasisTools._BASIS_FORMAT.cTFETC2;
+        }
+        return format;
+    }
+
+    /**
+     * Get the internal format to be passed to texImage2D corrisponding to the .basis format value
```
Comment (deltakosh) [inline src/Misc/basis.ts]: typo

File: Playground/index-local.html (hunk 2 above)
Comment (TrevorDev) [reply Playground/index-local.html]: I removed it from here to load the script using js like draco seems to work

File: src/Engines/engine.ts [hunk 7]
```diff
@@ -4975,6 +4975,30 @@ export class Engine {
         this._gl.compressedTexImage2D(target, lod, internalFormat, width, height, 0, <DataView>data);
     }
 
+    /** @hidden */
+    public _uploadDataToTextureAndResize(texture: InternalTexture, imageData: ArrayBufferView, width: number, height: number, babylonInternalFormat?: number) {
```
Comment (deltakosh) [inline src/Engines/engine.ts]: can we factorize it with the other usage in the engine (the one with IMG) ?

**Comment by deltakosh** [comment]
Only doc check is missing

**Comment by deltakosh** [comment]
still a doc issue :)

**Pull request merged**